6.x/custom-validation-rules.md

```diff
@@ -0,0 +1,67 @@
+# Custom Validation Rules
```

Comment:
Ok so after going through all the rules, I think I've spotted some inconsistencies in the API, which make these difficult to understand/remember/use. Here's a rundown of the current API:

```php
'avatar'      => ValidUpload::fileRules()->attributeRules(),
'attachments' => ValidUploadMultiple::fileRules()->arrayRules(),
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                'title' => 'required',
                'image' => 'required'
            ]),
```

This is... all over the place. We need to find a way to name rule methods in a consistent way.

My questions are:
(1) In `ValidUpload` what's `attributeRules()`? I don't understand.
(2) Does `ValidUploadMultiple->arrayRules()` do the same as  `ValidRepeatable::fieldRules()`? In that case we need to find some common names.
(3) To make the API consistent across custom validation rules, what do you think about:
- (A) having `fieldRules()` on ALL custom validations, which validates the big field; all custom validation rules will need that part, right?
- (B) having `itemRules()` on ALL custom validators that have sub-items; the difference being, `ValidUploadMultiple::itemRules()` will validate the files (those are the items) and `ValidRepeatable::itemRules()` will validate the sub-items/entries (those are the items).

After the above, the API would become:
```php
'avatar'      => ValidUpload::fieldRules(),
'attachments' => ValidUploadMultiple::fieldRules()->itemRules(),
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                'title' => 'required',
                'image' => 'required'
            ]),
```

Does the above make sense and cover everything? Let me know.

Reply:
TODO:
- [ ] add static `fieldRules()` to `ValidUpload`
- [ ] change docs to reflect recommended syntax:

```php
'avatar'      => ValidUpload::fieldRules()->fileRules(),
'attachments' => ValidUploadMultiple::fieldRules()->fileRules(), // alias: fileRules() == itemRules()
'gallery'     => ValidRepeatable::fieldRules()->itemRules([
                    'title' => 'required',
                    'image' => 'required',
                ]),
```